contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,213 @@ The central manager will poll each worker for their discovered volumes and creat
+4. Update the capacity in StorageClass object if there is a difference
+
+**Note:** For the alpha phase, thinly-provisioned VGs (overcommit) is not supported,
+We'll consider it in a subsequent phase.
+
+Though a list of VGs can be configured as storage source, it's not recommend to
+pass in more than one VG. We may suffer from some edge-cases:
+
+Suppose we have two VGs on a node for provisioning, VG1 and VG2, and they each has 10GB available,
+then the reported capacity would be 20GB, but a pod with storage request of 15GB could not
+be satisfied indeed.
+
+##### Change in Populator
+
+In addition to PV, the `populator` will also be extended to watch StorageClasses and PVC objects of all namespaces.
+
+It stores latest StorageClass objects in cache, they are used:
+
+* To reconcile capacity, the StorageManager need to compare actual capacity with that recorded
+in StorageClass objects
+* During volume creation, some opitions (e.g. fsType for `mkfs`) are specified in StorageClasses
+as Parameters
+
+And for PVCs, when the `populator` finds:
+
+1. An unbound PVC is annotated with `annSelectedNode`, and its node name match the value
+of the annotation
+2. Value of the `annStorageProvisioner` on the PVC matches the name of the provisioner
+
+It then trigger volume provsioning, detailed in the volume creation section.
+
+For these purposes, the following permissions are required:
+
+* list/get PersistentVolumeClaims
+* list/get/update StorageClasses
+
+##### Volume Creation
+
+During start-up of the provisioner, a new channel `provisionChan` will be created,
+and passed into both `populator` and `storageManager`:
+
+```
+provisionChan chan *v1.PersistentVolumeClaim
+```
+When it finds a PVC is in need of privsioning, as desicribed above,
+the `populator` will pass it into the channel to trigger volume privisioning;
+`StorageManager` will be the consumer, it will watch the channel and
+manage provisioning when new input received, that is:
+
+1. Create volumes behind a PV accordingly
+2. Create a PV object that is pre-bound to the PVC
+
+For LVM, the proposed workflow is:
+
+1. Check if there is already a PV pre-bound to the PVC, if so, return directly
+(This can happen when a volume has already been provisioned for a PVC,
+but has not been bound to the PVC by the controller yet).
+2. Check if Storage Class of the PVC is in record of its storageClass map.
+If not, return error.
+3. Check if the class is available for dynamic provisioning (i.e. has an input of `storageSources`).
+If not, return error.
+4. Check if there is a VG under the class, whose `Free Size` can satisfy the
+capacity requested in the PVC. If not, return error; otherwise, choose a VG for provisioning.
+5. Create `logical volume (LV)` object in the chosen VG.
+6. Initialize its file system with `mkfs`. FsType and other potential options
+for FS creation can be specified in StorageClass as Parameters
+7. Create a PV API object that is pre-bound to the PVC, and the path

```

Comment:
What is the path actually going to be?  Do we need the user to pass in a path where we can safely provision the lvs?

Reply:
I think LVM will create lvs under the path of vg automatically, so the path is determined once a vg is given.
And we just pass the path of the lv into the persistentvolume after that.